File: web-jpa/src/main/java/com/keunwon/jpashop/service/OrderService.java
package com.keunwon.jpashop.service;

import com.keunwon.jpashop.domain.Delivery;
import com.keunwon.jpashop.domain.Member;
import com.keunwon.jpashop.domain.Order;
import com.keunwon.jpashop.domain.OrderItem;
import com.keunwon.jpashop.domain.item.Item;
import com.keunwon.jpashop.repository.ItemRepository;
import com.keunwon.jpashop.repository.MemberRepository;
import com.keunwon.jpashop.repository.OrderRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@RequiredArgsConstructor
@Transactional(readOnly = true)
@Service
public class OrderService {

    private final OrderRepository orderRepository;
    private final MemberRepository memberRepository;
    private final ItemRepository itemRepository;

    @Transactional
    public Long order(Long memberId, Long itemId, int count) {
        Member member = memberRepository.findOne(memberId);
        Item item = itemRepository.findOne(itemId);

        Delivery delivery = new Delivery();
        delivery.setAddress(member.getAddress());

        OrderItem orderItem = OrderItem.createOrderItem(item, item.getPrice(), count);
        Order order = Order.createOrder(member, delivery, orderItem);

        orderRepository.save(order);
        return order.getId();
    }

    @Transactional
    public void cancelOrder(Long orderId) {
        Order order = orderRepository.findOne(orderId);
        order.cancel();
    }

    //public List<Order> findOrders(OrderSearch orderSearch) {}
}

File: spring/Spring-JWT/domains/domain-food/src/main/kotlin/com/github/keunwon/food/order/service/OrderService.kt
package com.github.keunwon.food.order.service

import com.github.keunwon.corejpa.getById
import com.github.keunwon.food.order.domain.OrderRepository
import com.github.keunwon.food.order.domain.OrderValidator
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class OrderService(
    private val orderRepository: OrderRepository,
    private val orderValidator: OrderValidator,
    private val orderMapper: OrderMapper,
) {
    @Transactional
    fun place(cart: Cart) {
        val order = orderMapper.mapFrom(cart).apply {
            place(orderValidator)
        }
        orderRepository.save(order)
    }

    @Transactional
    fun payOrder(orderId: Long) {
        val order = orderRepository.getById(orderId).apply { delivered() }
        orderRepository.save(order)
    }

    @Transactional
    fun deliverOrder(orderId: Long) {
        val order = orderRepository.getById(orderId).apply { delivered() }
        orderRepository.save(order)
    }
}
